lib/ts/recipe/multifactorauth/recipeImplementation.ts

```diff
@@ -0,0 +1,245 @@
+            //     ["thirdparty"],
+            // ];
+            // const setUpGroups = Array.from(
+            //     new Set(factorsSetUpForUser.map((id) => factorGroups.find((f) => f.includes(id)) || [id]))
+            // );
+
+            // const completedGroups = setUpGroups.filter((group) => group.some((id) => claimVal.c[id] !== undefined));
+
+            // // If the user completed every factor they could
+            // if (setUpGroups.length === completedGroups.length) {
+            //     logDebugMessage(
+            //         `isAllowedToSetupFactor ${factorId}: true because the user completed all factors they have set up and this is required`
+            //     );
+            //     return true;
+            // }
+
+            // return completedGroups.length >= 2;
+
+            const validator: SessionClaimValidator = {
+                id: MultiFactorAuthClaim.key,
+                claim: MultiFactorAuthClaim,
+                shouldRefetch: () => false,
+                validate: async (payload) => {
+                    const claimVal = MultiFactorAuthClaim.getValueFromPayload(payload);
+                    if (!claimVal) {
+                        throw new Error("should never happen");
+                    }
+
+                    const setOfUnsatisfiedFactors = MultiFactorAuthClaim.getNextSetOfUnsatisfiedFactors(
+                        claimVal.c,
+                        mfaRequirementsForAuth
+                    );
+
+                    if (setOfUnsatisfiedFactors.some((id) => factorsSetUpForUser.includes(id))) {
+                        logDebugMessage(
+                            `isAllowedToSetupFactor ${factorId}: false because there are items already set up in the next set of unsatisfied factors: ${setOfUnsatisfiedFactors.join(
+                                ", "
+                            )}`
+                        );
+                        return {
+                            isValid: false,
+                            reason: "Does not satisfy MFA requirements",
+                        };
+                    }
+
+                    logDebugMessage(
+                        `isAllowedToSetupFactor ${factorId}: true because the next set of unsatisfied factors is ${
+                            setOfUnsatisfiedFactors.length === 0 ? "empty" : "cannot be completed otherwise"
+                        }`
+                    );
+                    return { isValid: true };
+                },
+            };
+
+            await session.assertClaims([validator], userContext);
+        },
+
+        markFactorAsCompleteInSession: async function (this: RecipeInterface, { session, factorId, userContext }) {
+            const currentValue = await session.getClaimValue(MultiFactorAuthClaim);
+            const completed = {
+                ...currentValue?.c,
+                [factorId]: Math.floor(Date.now() / 1000),
+            };
+            const tenantId = session.getTenantId();
```

Comment:
instead of doing all the below.. just call setClaimValue with the new completed object, and then call session.fetchAndSetClaim(MultiFactorAuthClaim)? Why duplicate code?

Reply:
Double check that what i am saying makes sense or no..

Reply:
yea, it makes sense. Updated.